### PR TITLE
cp2k: migrate from homebrew/science

### PR DESCRIPTION
**Pull request by fxcoudert**
`depends_on "fftw" => "with-fortran"` is a pain, but vendoring fftw is ugly…

### CODE REVIEW

**Comment by fxcoudert** [comment]
Needs #21929

**Comment by fxcoudert** [comment]
MPI test fails on High Sierra CI because it cannot allocate 4 nodes (like all other MPI tests fail). It passes locally.

File: Formula/cp2k.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  needs :openmp
+
+  resource "libint" do
+    url "https://downloads.sourceforge.net/project/libint/v1-releases/libint-1.1.5.tar.gz"
+    sha256 "31d7dd553c7b1a773863fcddc15ba9358bdcc58f5962c9fcee1cd24f309c4198"
+  end
+
+  def install
+    resource("libint").stage do
+      system "./configure", "--prefix=#{prefix}/dependencies"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: libexec

File: Formula/cp2k.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      system "make"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: is the separate make step needed?

File: Formula/cp2k.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    # OpenMP versions link to specific fftw3 library
+    inreplace Dir["arch/Darwin-IntelMacintosh-gfortran.*smp"],
+              "-lfftw3", "-lfftw3 -lfftw3_threads"
+
+    # Now we build
+    cd "makefiles" do
+      %w[sopt ssmp popt psmp].each do |exe|
+        system "make", "ARCH=Darwin-IntelMacintosh-gfortran", "VERSION=#{exe}"
+        bin.install "../exe/Darwin-IntelMacintosh-gfortran/cp2k.#{exe}"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: have you requested a `make install` target?

File: Formula/cp2k.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        bin.install "../exe/Darwin-IntelMacintosh-gfortran/cp2k_shell.#{exe}"
+      end
+    end
+
+    # Install test file
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: comment is self evident

File: Formula/cp2k.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def install
+    resource("libint").stage do
+      system "./configure", "--prefix=#{prefix}/dependencies"
+      system "make"
+      system "make", "install"
+    end
+
+    fcflags = "-I#{Formula["libxc"].opt_prefix}/include -I#{Formula["fftw"].opt_prefix}/include -I#{prefix}/dependencies/include"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: See https://github.com/Homebrew/homebrew-core/blob/2c456f7019c2054dcf1f84af44f37aa3acedbbeb/Formula/osquery.rb#L178-L186 for how to format this kind of thing

Comment (ilovezfs) [reply Formula/cp2k.rb]: also use opt_include

File: Formula/cp2k.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    libs = "-L#{Formula["libxc"].opt_prefix}/lib -lxcf90 -lxc -L#{prefix}/dependencies/lib -lderiv -lint -L#{Formula["fftw"].opt_prefix}/lib -lfftw3"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: opt_lib

File: Formula/cp2k.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
+class Cp2k < Formula
+  desc "Quantum chemistry and solid state physics software package"
+  homepage "https://www.cp2k.org"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: trailing slash

File: Formula/cp2k.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      system "make"
+      system "make", "install"
+    end
+
+    fcflags = "-I#{Formula["libxc"].opt_prefix}/include -I#{Formula["fftw"].opt_prefix}/include -I#{prefix}/dependencies/include"
+    libs = "-L#{Formula["libxc"].opt_prefix}/lib -lxcf90 -lxc -L#{prefix}/dependencies/lib -lderiv -lint -L#{Formula["fftw"].opt_prefix}/lib -lfftw3"
+
+    # CP2K configuration is done through editing of arch files
+    inreplace Dir["arch/Darwin-IntelMacintosh-gfortran.*"].each do |s|
+      s.gsub! /DFLAGS *=/, "DFLAGS = -D__LIBXC -D__FFTW3 -D__LIBINT"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: can we use `change_make_var!`?

File: Formula/cp2k.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    end
+
+    # Install test file
+    (pkgshare/"tests").install "tests/Fist/water512.inp"
+  end
+
+  test do
+    system "#{bin}/cp2k.sopt", "#{pkgshare}/tests/water512.inp"
+    system "#{bin}/cp2k.ssmp", "#{pkgshare}/tests/water512.inp"
+    system "mpirun", "-np", "4", "#{bin}/cp2k.popt", "#{pkgshare}/tests/water512.inp"
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: removing -np 4 should fix the test on the mac pros

File: Formula/cp2k.rb (hunk 3 above)
Comment (fxcoudert) [reply Formula/cp2k.rb]: Yes, and it was declined:

```
install:
	@echo "The CP2K executable is $(PROG)"
```

File: Formula/cp2k.rb (hunk 8 above)
Comment (fxcoudert) [reply Formula/cp2k.rb]: Nope, because we want to preserve the existing content.

File: Formula/cp2k.rb (hunk 1 above)
Comment (ilovezfs) [reply Formula/cp2k.rb]: are any of
"--disable-debug",
"--disable-dependency-tracking",
"--disable-silent-rules",

valid?

Comment (fxcoudert) [reply Formula/cp2k.rb]: `--disable-debug` is the only valid one, and it's the default

File: Formula/cp2k.rb [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    resource("libint").stage do
+      system "./configure", "--prefix=#{libexec}"
+      system "make", "install"
+    end
+
+    fcflags = %W[
+      -I#{Formula["libxc"].opt_include}
+      -I#{Formula["fftw"].opt_include}
+      -I#{libexec}/include
+    ]
```
Comment (ilovezfs) [inline Formula/cp2k.rb]: blank line under here

File: Formula/cp2k.rb (hunk 3 above)
Comment (ilovezfs) [reply Formula/cp2k.rb]: #sad

**Pull request closed without merging**